distributed/scheduler.py

```diff
@@ -2523,11 +2523,27 @@ async def add_client(self, comm, client=None, versions=None):
             bcomm.start(comm)
             self.client_comms[client] = bcomm
             msg = {"op": "stream-start"}
+            scheduler_versions = version_module.get_versions()
             version_warning = version_module.error_message(
-                version_module.get_versions(),
+                scheduler_versions,
                 {w: ws.versions for w, ws in self.workers.items()},
                 versions,
             )
+            # NOTE: whether or not this throws or merely warns should be
+            # configurable. It would also be great to have an error on
+            # a major version mismatch, which could cause errors.
+            client_versions = versions
+            if client_versions:
+                from .protocol import compression
+
+                if not client_versions.get(compression.default_compression, None):
+                    version_warning = version_warning or ""
+                    version_warning += """
+
+DANGER! The scheduler's default compression method is not available
+in this client! You may not be able to receive large result sets!
+                    """
+
```

Comment:
I like the idea of calling out specific library versions that may be particularly concerning.

If we do this though then we might want to be more comprehensive.  The compression mismatch can cause a serious issue if any two components (clients, schedulers, or workers) disagree with each other, not just the client/scheduler relationship.  In an ideal world two peers would do a little handshake in order to determine which compression to use.  We're not there yet today though.

Short term though, if we want to put this somewhere I wonder if we want it to be in `versions.py::error_message`.  I haven't actually dug into this yet, so please don't rush and change a bunch of things.  This is mostly a question.

Reply:
I'd be totally fine putting it in `versions::error_message` for the moment, and rely on the `Invalid compression method` message for the user to recognize the error even if they aren't paying attention to the warning. I'd like to note that while I understand the warning _is_ shown by default, it's really easy to skim over, especially when most of the mismatches are likely to be minor version issues that probably won't cause problems. So I'd ideally like to have some way to fail fast on connecting, if an error is likely.

 I'll hold off on making changes for now, though -- thank you for taking a look!